### DISCUSSION TITLE
[vcpkg upgrade] Please make vcpkg remember what packages are still to be rebuilt, when one of them fails during an upgrade

### DISCUSSION THREAD

**Discussion by dannypike** [microsoft/vcpkg#26638] [Ideas]
When I use ```vcpkg upgrade --no-dry-run```, I'm always nervous because one of the updated ports may fail the build. At that point ```vcpkg``` will have "forgotten" all of the ports that it removed while trying to upgrade everything.

So I'm left with a half-broken system and I have to remember all of the ports that I need to install in order to make my various projects build, which can be frustrating and time consuming.

If you can automate the recovery, that would be wonderful but, if not, would it be possible for ```vcpkg``` to list which ports have not yet been rebuilt, just before it exits with a failure? I can then restart those ones only, saving a huge amount of time.

**Comment by SKNo** [comment]
Hi
I use `vcpkg upgrade --no-dry-run --keep-going`
In the end, it gives a summary of successful and unsuccessful builds. It also forces to continue build process even if one (or few) of the packages or triplets fails to be built.

**Comment by dannypike** [comment]
Thank you. I didn't know about that option; it's not in the help text or the README. A quick search through the docs only finds it buried in the change list for one of the updates to the `install` command.

I suppose that's slightly better than being "...display[ed] in the bottom of a locked filing cabinet stuck in a disused lavatory with a sign on the door saying 'Beware of The Leopard'", but not much. :)

**Comment by wrobelda** [comment]
> So I'm left with a half-broken system and I have to remember all of the ports that I need to install in order to make my various projects build, which can be frustrating and time consuming.

This sounds like you could take advantage of the manifest mode: https://vcpkg.io/en/docs/users/manifests.html

**Comment by dannypike** [comment]
@wrobelda Thank you for the suggestion but the problem is that I use `vcpkg` for several projects, both mine and also others'. So manifest mode won't really work.

What I want is for `vcpkg` to record the current setup in some way so that, if the upgrade fails, I will be able to restart it, e.g. when the failing project is fixed.

Using `--keep-going` is a significant help, but I would prefer `vcpkg` to not forget what it is trying to `upgrade` when that upgrade fails partway through. It doesn't need to be over-engineered and 100% fail safe. We're all developers here and (hopefully!) won't do anything too stupid.

For example, I'd be happy to use a different "mode" for restarting the upgrade using a log file that contains a list of packages that it last removed, e.g.:

```
vcpkg upgrade --restart-last --no-dry-run --keep-going
```

would not need to do anything new or clever; it would only rebuild all of the packages that the last call to:

```
vcpkg upgrade --no-dry-run --keep-going
```

attempted, by using some sort of log file that it recorded at the start of the previous attempt. Successful packages would be skipped, in the same way that `vcpkg` always works.

Whenever I do a full upgrade again:
```
vcpkg upgrade --no-dry-run --keep-going
```
`vcpkg` would then overwrite the aforesaid log file with a new list of installed packages that it is removing and the previous information would be lost. That would be fine and solve my problem - which is remembering dependencies for loads of different projects.
